moshi/src/main/java/com/squareup/moshi/ObjectJsonWriter.java

```diff
@@ -131,7 +140,20 @@ public Object root() {
   }
 
   @Override public JsonWriter value(double value) throws IOException {
-    return value(Double.valueOf(value));
+    long longValue = (long) value;
+    if (Double.compare(longValue, value) == 0) {
+      return value(longValue);
```

Comment:
I'm inclined to agree here. Seems like if you have a double, even one without a floating part, then we should retain it as a double.